### PR TITLE
Implement beacon sync

### LINKED ISSUE #2622: Implement tortoise beacon sync
## Description

Tortoise beacon sync should be implemented according to https://github.com/spacemeshos/SMIPS/issues/52

## Affected code

Tortoise beacon, sync

This issue appears in commit hash: 
`4b13d81196bb17698fba6312f81c0e54b8e73a76`

Related files (optionally with line numbers): 
TBD

### PR DESCRIPTION
**Pull request by nkryuchkov**
## Motivation

Closes #2622

## Changes

- Implement beacon sync

## TODO

- [ ] Finish the implementation
- [ ] Test changes and document test plan

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -281,13 +292,30 @@ func (s *Syncer) synchronize(ctx context.Context) bool {
...
+		layer, err := s.syncLayer(ctx, layerID)
+		if err != nil {
 			logger.With().Error("failed to sync to layer", layerID, log.Err(err))
 			return false
 		}
 
+		if layerID.OrdinalInEpoch() < s.conf.TBCalculationLayers {
+			s.processTortoiseBeacons(ctx, layer)
+		} else {
+			epoch := layerID.GetEpoch()
```
Comment (antonlerner) [inline syncer/syncer.go:313]: should this be done for each layer as of `s.conf.TBCalculationLayers` ?

File: syncer/syncer.go [hunk 2]
```diff
@@ -281,13 +292,30 @@ func (s *Syncer) synchronize(ctx context.Context) bool {
...
+		}
+
+		layer, err := s.syncLayer(ctx, layerID)
+		if err != nil {
 			logger.With().Error("failed to sync to layer", layerID, log.Err(err))
 			return false
 		}
 
+		if layerID.OrdinalInEpoch() < s.conf.TBCalculationLayers {
+			s.processTortoiseBeacons(ctx, layer)
```
Comment (antonlerner) [inline syncer/syncer.go:311]: are we supposed to process TBs only for `s.conf.TBCalculationLayers` layers?

File: syncer/syncer.go [hunk 3]
```diff
@@ -417,6 +448,25 @@ func (s *Syncer) getATXs(ctx context.Context, layerID types.LayerID) error {
 	return nil
 }
 
+func (s *Syncer) processTortoiseBeacons(ctx context.Context, layer *types.Layer) {
+	for _, b := range layer.Blocks() {
+		if _, ok := s.seenTB[b.MinerID().String()]; ok {
+			// TODO(nkryuchkov): invalid block, miner attempted to include beacon in two blocks within the same epoch
```
Comment (antonlerner) [inline syncer/syncer.go:454]: add log for this

File: syncer/syncer.go [hunk 4]
```diff
@@ -26,15 +27,17 @@ type layerFetcher interface {
 	PollLayerHash(ctx context.Context, layerID types.LayerID) chan layerfetcher.LayerHashResult
 	PollLayerBlocks(ctx context.Context, layerID types.LayerID, hashes map[types.Hash32][]peers.Peer) chan layerfetcher.LayerPromiseResult
 	GetEpochATXs(ctx context.Context, id types.EpochID) error
-	GetTortoiseBeacon(ctx context.Context, id types.EpochID) error
+	GetTortoiseBeacon(ctx context.Context, id types.EpochID) error // TODO(nkryuchkov): remove
```
Comment (antonlerner) [inline syncer/syncer.go]: fetcher shouldn't contain these methods, as it is not related to any fetching of data, where is tbDB created? where is it used except for the fetcher?

File: syncer/syncer.go [hunk 5]
```diff
@@ -417,6 +448,25 @@ func (s *Syncer) getATXs(ctx context.Context, layerID types.LayerID) error {
...
+			// TODO(nkryuchkov): invalid block, miner attempted to include beacon in two blocks within the same epoch
+		}
+
+		s.seenTB[b.MinerID().String()] = struct{}{}
+
+		tb := b.EligibilityProof.TortoiseBeacon
+		tbString := string(tb)
+
+		s.tbCount[tbString]++
+		if s.tbCount[tbString] > s.mostUsedTBCount {
```
Comment (antonlerner) [inline syncer/syncer.go]: how is the first TB defined? where does the value come from?

File: syncer/syncer.go (hunk 1 above)
Comment (nkryuchkov) [reply syncer/syncer.go:313]: No, only one. Fixed it

File: syncer/syncer.go (hunk 2 above)
Comment (nkryuchkov) [reply syncer/syncer.go:311]: yes, it's `K` from the beacon sync discussion

File: syncer/syncer.go (hunk 4 above)
Comment (nkryuchkov) [reply syncer/syncer.go]: `tbDB` is created in `node.go`, do you think it should be passed to syncer?

File: syncer/syncer.go (hunk 5 above)
Comment (nkryuchkov) [reply syncer/syncer.go]: Do you mean the initial value of `mostUsedTBCount`?

**Comment by nkryuchkov** [comment]
@countvonzero Should we close it as it will be overridden with your PR(s)?

**Comment by countvonzero** [comment]
@nkryuchkov yes. this is obsolete.

**Pull request closed without merging**